### PR TITLE
Fixing clipping

### PR DESCRIPTION
**Pull request by zvallisVividQ**
plus other changes:
- separating renderer
- more unit tests
- detection if cursor goes outside program window

### CODE REVIEW

File: PaintTests/WinTests.cpp [hunk 1]
```diff
@@ -609,7 +609,7 @@ namespace PaintTests
 			Renderer* rendererPtr = &dummyRenderer;
 			uint8_t ColA = 0;
```
Comment (lordGoldemort) [inline PaintTests/WinTests.cpp]: Should be camelCase, not PascalCase
**SHOULD**

File: gfx/GFX_Renderer.h [hunk 2]
```diff
@@ -36,9 +36,9 @@ namespace gfx
 		void renderPresent() const;
```
Comment (lordGoldemort) [inline gfx/GFX_Renderer.h]: Preferred structure to keep gfx free from SDL contamination

project gfx
```
    class Renderer {
        virtual void renderBox(..) = 0;
        virtual void renderText(…) = 0;
        // etc

    protected:
        Renderer() = default;
    };

```
project paint 
```
    class SDLRenderer : public gfx::Renderer {
    public:
        SDLRenderer(SDL_Renderer * sdl)
        : sdl_(sdl)
       {
       }
       
       virtual void renderBox(...) override;
       virtual void renderBox(...) override;
       virtual void renderBox(...) override;

    private:
        SDL_Renderer * sdlRenderer_;
    };
    

```
**SHOULD**

File: paint/PAINT_DrawWindow.cpp [hunk 3]
```diff
@@ -192,29 +204,24 @@ void DrawWindow::clearWindow() const
 //using Cohen-Sutherland clipping algorithm
 std::vector<win::Coords> DrawWindow::clippingHandler(win::Coords pStart, win::Coords pEnd) const
```
Comment (lordGoldemort) [inline paint/PAINT_DrawWindow.cpp]: Prefer this to be a utility function in gfx

```
namespace gfx {
    std::vector<Point> clipLineToRect(Rect const & r, Point const & start, Point const & end);
}
```
            
(Would need to move Coord from win -> gfx OR create a Point type in gfx and a function to convert (hopefully just a cast!))
**SHOULD**

File: paint/PAINT_DrawWindow.h [hunk 4]
```diff
@@ -38,9 +38,10 @@ namespace paint
 		DrawWindow& operator=(const DrawWindow& that) = delete;
 		DrawWindow& operator=(DrawWindow&& that) = delete;
 
+		bool mouseEnter(bool clicked = false) override;
```
Comment (lordGoldemort) [inline paint/PAINT_DrawWindow.h:40]: Just a comment. In future we might have the possibility of using the right mouse button to draw background colours whilst the left draws foreground. So a boolean woulndn't suffice here.

File: paint/PAINT_Eraser.h [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 namespace paint
 {
-	class Eraser final :
+	class Eraser :
```
Comment (lordGoldemort) [inline paint/PAINT_Eraser.h:6]: Why remove `final`?

File: paint/PAINT_StatusBarWindow.h [hunk 6]
```diff
@@ -9,7 +9,7 @@ namespace paint
 	{
 	public:
 		StatusBarWindow() = delete;
-		StatusBarWindow(gfx::Renderer* renderer, gfx::Rectangle const& rect, const char* name, const int drawWinX, const int drawWinY);
+		StatusBarWindow(gfx::Renderer* renderer, gfx::Rectangle const& rect, const char* name, int drawWinX, int drawWinY);
```
Comment (lordGoldemort) [inline paint/PAINT_StatusBarWindow.h]: Don't construct with the renderer, just pass it into draw. Then there's no scope for doing renderer-related things in non-renderer-related contexts.
**MUST**

File: win/WIN_Slider.cpp [hunk 7]
```diff
@@ -183,7 +184,7 @@ bool Slider::mouseButtonUp(MouseButton button)
 		int xMouse = 0;
 		int yMouse = 0;
 		
-		SDL_GetMouseState(&xMouse, &yMouse);
+		getRenderer()->getMouseState(xMouse, yMouse);
```
Comment (lordGoldemort) [inline win/WIN_Slider.cpp]: Should probably just pass these in as parameters. This violates SRP (Single Responsibility Principle) - a Renderer should just be for rendering, not for getting peripheral state).
**MUST**

File: win/WIN_Utils.h [hunk 8]
```diff
@@ -8,7 +8,7 @@ namespace win
 {
 	namespace utils
 	{
-		bool filterNumerical(const char c);
+		bool filterNumerical(char c);
 
 		int findOutcode(gfx::Rectangle rect, int x, int y);
```
Comment (lordGoldemort) [inline win/WIN_Utils.h]: Should be part of gfx.
**MUST**

**Review by lordGoldemort — APPROVED** [review]
ReSharper reports:

Solution paint.sln
    Project paint
      paint\PAINT_ButtonFunctions.cpp:18 Typo in comment: "cppcoreguidelines"
      paint\PAINT_ButtonFunctions.cpp:33 Typo in comment: "cppcoreguidelines"
      paint\PAINT_ButtonFunctions.cpp:78 Typo in comment: "Rerender"
      paint\PAINT_ColourDisplay.cpp:102 Typo in identifier: "cpick"
      paint\PAINT_ColourPicker.cpp:175 Typo in identifier: "Rerender"
      paint\PAINT_ColourPicker.h:38 Typo in identifier: "Rerender"
      paint\PAINT_DrawWindow.cpp:28 Typo in comment: "PIXELFORMAT"
      paint\PAINT_DrawWindow.cpp:28 Typo in comment: "TEXTUREACCESS"
      paint\PAINT_DrawWindow.cpp:169 Typo in comment: "cout"
      paint\PAINT_DrawWindow.cpp:171 Typo in comment: "cout"
      paint\PAINT_DrawWindow.cpp:192 Typo in identifier: "Rerender"
      paint\PAINT_DrawWindow.h:68 Typo in identifier: "Rerender"
      paint\PAINT_Program.cpp:91 Typo in comment: "rerender"
      paint\PAINT_Program.cpp:92 Typo in identifier: "rerender"
      paint\PAINT_Screen.cpp:121 Typo in identifier: "Rerender"
      paint\PAINT_Screen.h:32 Typo in identifier: "Rerender"
      paint\PAINT_Utils.cpp:36 Typo in comment: "cppcoreguidelines"

File: PaintTests/WinTests.cpp (hunk 1 above)
Comment (zvallisVividQ) [reply PaintTests/WinTests.cpp]: done

File: gfx/GFX_Renderer.h (hunk 2 above)
Comment (zvallisVividQ) [reply gfx/GFX_Renderer.h]: done

File: paint/PAINT_DrawWindow.cpp (hunk 3 above)
Comment (zvallisVividQ) [reply paint/PAINT_DrawWindow.cpp]: done, and moved Coord to gfx too

File: paint/PAINT_Eraser.h (hunk 5 above)
Comment (zvallisVividQ) [reply paint/PAINT_Eraser.h:6]: Resharper complained, I've added it back

File: paint/PAINT_StatusBarWindow.h (hunk 6 above)
Comment (zvallisVividQ) [reply paint/PAINT_StatusBarWindow.h]: removed renderer_ pointer for windows where it's not necessary, and added draw overload with renderer

File: win/WIN_Slider.cpp (hunk 7 above)
Comment (zvallisVividQ) [reply win/WIN_Slider.cpp]: I put the mouse state function here to separate SDL from Slider (and other classes that rely on mouse states), but there's nothing that requires it be in Renderer. So I've moved getMouseState out of Renderer into win::utils.

File: win/WIN_Utils.h (hunk 8 above)
Comment (zvallisVividQ) [reply win/WIN_Utils.h]: Done, outcode stuff is now in gfx. Win utils contains only filterNumerical and getMouseState now

**Pull request merged**